mlir-sycl/lib/Dialect/SYCL/Transforms/ConstantPropagation.cpp

```diff
@@ -91,12 +112,630 @@ class ConstantArithArg : public ConstantExplicitArg {
+      [](const auto *ptr) { return ptr->printDebugMessage(); });
+}
+
+void ConstantImplicitArgBase::propagate(FunctionOpInterface func) {
+  LLVM_DEBUG(printDebugMessage());
+  SymbolTableCollection symbolTable;
+  propagate(func, symbolTable, isRecursive());
+}
+
+bool ConstantImplicitArgBase::needsChanges(FunctionOpInterface func,
+                                           SymbolTableCollection &symbolTable,
+                                           bool recursive) const {
+  // Declaration
+  if (func.isExternal())
+    return false;
+  return func
+      .walk([&](Operation *op) {
+        if (succeeded(match(op)))
+          return WalkResult::interrupt();
+        if (!recursive)
+          return WalkResult::advance();
+        auto call = dyn_cast<CallOpInterface>(op);
+        if (!call)
+          return WalkResult::advance();
+        SymbolRefAttr callable =
+            call.getCallableForCallee().dyn_cast<SymbolRefAttr>();
+        if (!callable)
+          return WalkResult::advance();
+        auto called = symbolTable.lookupNearestSymbolFrom<FunctionOpInterface>(
+            call, callable);
+        assert(called && "Function not found");
+        return needsChanges(called, symbolTable, recursive)
+                   ? WalkResult::interrupt()
+                   : WalkResult::advance();
+      })
+      .wasInterrupted();
+}
+
+void ConstantImplicitArgBase::propagate(FunctionOpInterface func,
+                                        SymbolTableCollection &symbolTable,
+                                        bool recursive, Twine callPath,
+                                        CallOpInterface call) {
+  LLVM_DEBUG(llvm::dbgs().indent(6)
+             << "Entering function '"
+             << static_cast<SymbolOpInterface>(func).getName() << "'\n");
+  // Declaration
+  if (func.isExternal())
+    return;
+  StringRef originalName = static_cast<SymbolOpInterface>(func).getName();
+  // Need to clone called functions to specialize
+  if (call) {
+    // Avoid cloning if no change will be performed.
+    // Only check whether func needs changes here to avoid iterating twice.
+    if (!needsChanges(func, symbolTable, recursive))
+      return;
+    func = cloneFunction(func, call, callPath);
+    LLVM_DEBUG(llvm::dbgs().indent(8)
+               << "Cloned into '"
+               << static_cast<SymbolOpInterface>(func).getName()
+               << "' to specialize\n");
+  }
+  OpBuilder builder(func.getFunctionBody());
+  func.walk([&](Operation *op) {
+    if (succeeded(match(op))) {
+      rewrite(op, builder);
+      recordHit();
+      return;
+    }
+    if (!recursive)
+      return;
+    auto call = dyn_cast<CallOpInterface>(op);
+    if (!call)
+      return;
+    SymbolRefAttr callable =
+        call.getCallableForCallee().dyn_cast<SymbolRefAttr>();
+    if (!callable)
+      return;
+    auto called = symbolTable.lookupNearestSymbolFrom<FunctionOpInterface>(
+        call, callable);
```

Comment:
Do you think is useful to have this section of code in a private function? so it can be reused.

Reply:
I think some of this code will need refactoring, but I'd rather do that after our deadline. This will probably need implementing our own "walker" struct to unify `needsChanges` and `propagate` code.